File: LeetCode/Java Solutions/Single Number III/single.java
class Solution {
    public int[] singleNumber(int[] nums) {
        // Step 1: XOR all numbers
        int xorResult = 0;
        for (int num : nums) {
            xorResult ^= num;
        }
        
        // Step 2: Find a differentiating bit (rightmost set bit)
        int differentiatingBit = xorResult & -xorResult;
        
        // Step 3: Partition the numbers and XOR within each partition
        int num1 = 0;
        int num2 = 0;
        for (int num : nums) {
            if ((num & differentiatingBit) != 0) {
                num1 ^= num;
            } else {
                num2 ^= num;
            }
        }
        
        return new int[]{num1, num2};
    }

    // Test cases
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] result1 = sol.singleNumber(new int[]{1, 2, 1, 3, 2, 5});
        System.out.println(Arrays.toString(result1));  // Output: [3, 5]

        int[] result2 = sol.singleNumber(new int[]{-1, 0});
        System.out.println(Arrays.toString(result2));  // Output: [-1, 0]

        int[] result3 = sol.singleNumber(new int[]{0, 1});
        System.out.println(Arrays.toString(result3));  // Output: [1, 0]
    }
}

File: LeetCode/Python Solutions/Single Number III/single.py
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        # Step 1: XOR all numbers
        xor_result = 0
        for num in nums:
            xor_result ^= num
        
        # Step 2: Find a differentiating bit (rightmost set bit)
        differentiating_bit = xor_result & -xor_result
        
        # Step 3: Partition the numbers and XOR within each partition
        num1 = 0
        num2 = 0
        for num in nums:
            if num & differentiating_bit:
                num1 ^= num
            else:
                num2 ^= num
        
        return [num1, num2]

# Test cases
sol = Solution()
print(sol.singleNumber([1, 2, 1, 3, 2, 5]))  # Output: [3, 5]
print(sol.singleNumber([-1, 0]))  # Output: [-1, 0]
print(sol.singleNumber([0, 1]))  # Output: [1, 0]
